network/index.md

```diff
@@ -1,103 +1,95 @@
+
+## Published ports
+
+By default, when you create or run a container using `docker create` or `docker run`,
+the container doesn't expose any of its ports to the outside world.
+To make a port available to services outside of Docker,
+or to Docker containers running on a different network,
+use the `--publish` or `-p` flag.
+This creates a firewall rule in the container,
+mapping a container port to a port on the Docker host to the outside world.
+Here are some examples:
+
+| Flag value                      | Description                                                                                                                                           |
+| ------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------- |
+| `-p 8080:80`                    | Map TCP port 80 in the container to port `8080` on the Docker host.                                                                                   |
+| `-p 192.168.1.100:8080:80`      | Map TCP port 80 in the container to port `8080` on the Docker host for connections to host IP `192.168.1.100`.                                        |
+| `-p 8080:80/udp`                | Map UDP port 80 in the container to port `8080` on the Docker host.                                                                                   |
+| `-p 8080:80/tcp -p 8080:80/udp` | Map TCP port 80 in the container to TCP port `8080` on the Docker host, and map UDP port `80` in the container to UDP port `8080` on the Docker host. |
+
+## IP address and hostname
+
+By default, the container gets an IP address for every Docker network it attaches to.
+A container receives an IP address out of the IP pool of the network it attaches to.
+The Docker daemon effectively acts as a DHCP server for each container.
+Each network also has a default subnet mask and gateway.
+
+When a container starts, it can only attach to a single network, using the `--network` flag.
+You can connect a running container to multiple networks using the `docker network connect` command.
+When you start a container using the `--network` flag,
+you can specify the IP address for the container on that network using the `--ip` or `--ip6` flags.
+
+When you connect an existing container to a different network using `docker network connect`,
+you can use the `--ip` or `--ip6` flags on that command
+to specify the container's IP address on the additional network.
+
+In the same way, a container's hostname defaults to be the container's ID in Docker.
+You can override the hostname using `--hostname`.
+When connecting to an existing network using `docker network connect`,
+you can use the `--alias` flag to specify an additional network alias for the container on that network.
+
+## DNS services
+
+By default, containers inherit the DNS settings of the host, as defined in the `/etc/resolv.conf` configuration file.
```

Comment:
I recall working on some of the "hashData" code a while back; https://github.com/moby/moby/blob/8d67d0c1a8d28ad4014444edc6b9975f9877b1d6/libnetwork/resolvconf/utils.go#L8-L14

Which I seem to recall, was used to detect whether the file was modified, and (MAYBE) was used to decide whether or not the `/etc/resolv.conf` could be updated, or should stay untouched (to prevent overwriting user's modifications) I really need to check on that though.

Reply:
The hash of `resolv.conf` is used here: https://github.com/moby/moby/blob/8d67d0c1a8d28ad4014444edc6b9975f9877b1d6/libnetwork/sandbox_dns_unix.go#L313-L317. This code only checks whether the container's `resolv.conf` has been edited by the end-user, and if not (re)writes it. AFAIK, it's run during live-restore and when `--link` is used.

For the upstream DNS server(s) used by embedded resolver, we have a `(*Sandbox).rebuildDNS()` which could do that. But its name is a bit misleading because it's only called by `(*Sandbox).startResolver()` and gated by a `!restore`. So, at this point _I think_ we never re-read host `resolv.conf` to update existing embedded resolver, even during a live restore. But I might be wrong, libnetwork being what it is.

Reply:
I see no action to take here for now. As stated earlier, we will do networking differences between CE/DD in a follow-up.